ext/IPC-Open3/lib/IPC/Open3.pm

```diff
@@ -257,6 +257,16 @@ sub _open3 {
 		untie *STDOUT;
 		untie *STDERR;
 
+		if ((fileno(STDIN)//-1) != 0) {
+                        eval { close(STDIN); };
```

Comment:
This is a huge hammer with global program effects at a distance and I don't think it's a good idea. This problem only occurs when STDIN or STDOUT (or STDERR) is closed, and I would consider that the bug in any such program - for the common task of daemonization, one should instead reopen them to `File::Spec->devnull`. Perhaps the documentation could mention this.

Reply:
To put it more simply, if STDIN is not fd 0, then this line is closing an arbitrary filehandle we don't own. I don't consider that acceptable behavior.

Reply:
This is happening in the _child_ process, which is about to `exec` another program.  If STDIN/STDOUT do not refer to UNIX fds 0/1 respectively, then there is _no way_ to set up the pipes correctly.

I consider it a serious defect that `open3` can fail because of what STDIN/STDOUT refer to in the parent process.

I would accept changing the patch to only open STDIN/STDOUT if they have no associated fd (meaning they're open for in-memory I/O), but I think that's wrong.

Your remarks about global effects at a distance are not relevant because as mentioned, we're about to exec a program anyway.

Reply:
Yes, the defect is that STDIN/STDOUT have been broken by the parent process.

Reply:
The global effect is relevant because closing the handle in the child process affects the underlying handle.

Reply:
Actually, another option would be to `fdopen` our own private handles to fd 0 and 1 respectively, and use those (rather than STDIN/STDOUT) to set up the pipes.

Basically, we need a way to get Perl to set up pipes using POSIX file descriptors 0 and 1 regardless of how the caller might have mucked about with STDIN and STDOUT.  Because ultimately it's the file descriptors that matter, not the Perl filehandles.

It's the weekend, but I'll look at this on Monday.

Reply:
Keep in mind also that IPC::Open3 does not always exec:

```
Note if you specify - as the command, in an analogous fashion to open(my $fh, "-|") the child process will just be the forked Perl process rather than an external command.
```